File: leetcode/0001-1000/0101-0200/0111-0120/0111-minimum-depth-of-binary-tree/MinimumDepthOfBinaryTree.java
import java.util.PriorityQueue;

class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode() {
  }

  TreeNode(int val) {
    this.val = val;
  }

  TreeNode(int val, TreeNode left, TreeNode right) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

class Solution {
  public int minDepth(TreeNode root) {
    if (root == null) {
      return 0;
    }

    var queue = new PriorityQueue<Object[]>((a, b) -> {
      var heightA = (Integer) a[1];
      var heightB = (Integer) b[1];

      return Integer.compare(heightA, heightB);
    });

    queue.offer(new Object[] { root, 1 });

    while (!queue.isEmpty()) {
      var objects = (Object[]) queue.poll();
      var node = (TreeNode) objects[0];
      var height = (Integer) objects[1];

      if (node.left == null && node.right == null) {
        return height;
      }

      if (node.left != null) {
        queue.offer(new Object[] { node.left, height + 1 });
      }

      if (node.right != null) {
        queue.offer(new Object[] { node.right, height + 1 });
      }
    }

    return 1;
  }
}

File: leetcode/0001-1000/0101-0200/0111-0120/0111-minimum-depth-of-binary-tree/MinimumDepthOfBinaryTree.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function minDepth(root: TreeNode | null): number {
  if (root == null) {
    return 0;
  }

  let queue: TreeNode[] = [root];
  let height = 0;

  while (queue.length > 0) {
    let size = queue.length;
    height++;

    for (let i = 0; i < size; i++) {
      let node = queue.shift()!;

      if (node.left == null && node.right == null) {
        return height;
      }

      if (node.left != null) {
        queue.push(node.left);
      }

      if (node.right != null) {
        queue.push(node.right);
      }
    }
  }

  return 0;
}
